external/js/cothority/src/network/connection.ts

```diff
@@ -191,67 +188,66 @@ export class WebSocketConnection implements IConnection {
     }
 
     /** @inheritdoc */
-    sendStream<T extends Message>(message: Message, reply: typeof Message,
-                                  onMessage: (data: T, ws: WebSocketAdapter) => void,
-                                  onClose: (code: number, reason: string) => void,
-                                  onError: (err: Error) => void): WebSocketAdapter {
+    sendStream<T extends Message>(message: Message, reply: typeof Message):
+        Observable<[T, WebSocketAdapter]> {
+        return new Observable((sub) => {
 
-        if (!message.$type) {
-            onError(new Error(`message "${message.constructor.name}" is not registered`));
-            return;
-        }
+            if (!message.$type) {
+                sub.error(new Error(`message "${message.constructor.name}" is not registered`));
+                return;
+            }
 
-        if (!reply.$type) {
-            onError(new Error(`message "${reply}" is not registered`));
-            return;
-        }
+            const path = this.getURL() + "/" + this.service + "/" + message.$type.name.replace(/.*\./, "");
+            Log.lvl4(`Socket: new WebSocket(${path})`);
+            const ws = factory(path);
+            const bytes = Buffer.from(message.$type.encode(message).finish());
+            const timer = setTimeout(() => ws.close(1000, "timeout"), this.timeout);
```

Comment:
It's the same preamble as `send`, consider putting it inside an helper. (it took so much stressfull minutes to try to wrap my head around the fact that my commit about URL was merged in master, but this code was not using it; basic humain mistake of matching `send` and `sendStream` but yeah, code duplication)

Reply:
good idea.